crates/ntx-builder/src/builder/server/mod.rs

```diff
@@ -0,0 +1,139 @@
+use std::sync::{Arc, Mutex};
+
+use miden_node_proto::{
+    generated::{
+        ntx_builder::api_server::Api,
+        requests::{
+            SubmitNetworkNotesRequest, UpdateNetworkNotesRequest, UpdateTransactionStatusRequest,
+        },
+        transaction::TransactionStatus,
+    },
+    try_convert,
+};
+use miden_objects::{
+    Digest,
+    note::{Note, Nullifier},
+};
+use state::NtxBuilderState;
+use tonic::{Request, Response, Status};
+use tracing::info;
+
+use crate::COMPONENT;
+
+mod state;
+
+#[derive(Debug)]
+pub struct NtxBuilderApi {
+    state: Arc<Mutex<NtxBuilderState>>,
+}
+
+impl NtxBuilderApi {
+    pub fn new(unconsumed_network_notes: Vec<Note>) -> Self {
+        let state = NtxBuilderState::new(unconsumed_network_notes);
+        Self { state: Arc::new(Mutex::new(state)) }
+    }
+
+    pub fn state(&self) -> Arc<Mutex<NtxBuilderState>> {
+        self.state.clone()
+    }
+}
+
+#[tonic::async_trait]
+impl Api for NtxBuilderApi {
+    async fn submit_network_notes(
+        &self,
+        request: Request<SubmitNetworkNotesRequest>,
+    ) -> Result<Response<()>, Status> {
+        let req = request.into_inner();
+        let tx_id = req.transaction_id.ok_or_else(|| {
+            Status::invalid_argument("transaction_id is required in SubmitNetworkNotesRequest")
+        })?;
+
+        info!(
+            target: COMPONENT,
+            tx_id = %tx_id,
+            note_count = req.note.len(),
+            "Received network notes"
+        );
```

Comment:
Might want to instrument this function. In which case could probably remove this event given its informational value is in the fields.